In [1]:
import torch
import gc
torch.cuda.empty_cache()
gc.collect()


0

In [2]:
from setup import setup_src_path
print(setup_src_path())
import data.processed as processed
import config.config as config
import utils.setup as setup
import utils.functions as fn
from importlib import reload



['/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/mlm', '/usr/lib/python38.zip', '/usr/lib/python3.8', '/usr/lib/python3.8/lib-dynload', '', '/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages', '/home/guest/Desktop/projects/third-experiments/domain_adaptation_project/modules']


2024-06-04 03:07:59.604368: I tensorflow/core/util/port.cc:110] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2024-06-04 03:07:59.637538: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX512F AVX512_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
2024-06-04 03:08:00.386178: W tensorflow/compiler/tf2tensorrt/utils/py_utils.cc:38] TF-TRT Warning: Could not find TensorRT


In [3]:
from datasets import load_from_disk,concatenate_datasets


adapter_name="mlm_union_F"
dataset = load_from_disk(f"{config.Config.DATASETS_SAVE_PATH}/datasets")



In [4]:
reload(config)

    
filtered_target = dataset['train'].filter(lambda example: example['genre'] == "fiction").shuffle(seed=42)
#filtered_target = shuffled_filtered_target.train_test_split(test_size=0.1)


filtered_test_target = dataset['validation_matched'].filter(lambda example: example['genre'] == "fiction")
train_target = filtered_target
test_target = filtered_test_target

 

Filter:   0%|          | 0/392702 [00:00<?, ? examples/s]

Filter:   0%|          | 0/9815 [00:00<?, ? examples/s]

In [5]:
from adapters import AutoAdapterModel,init
from tqdm import tqdm

from transformers import AutoConfig, AutoModelForMaskedLM

mdlcfg = AutoConfig.from_pretrained(
    config.Config.MODEL_NAME,
 
)
model = AutoModelForMaskedLM.from_pretrained(
    config.Config.MODEL_NAME,
)
init(model)
reload(fn)
fn.print_trainable_parameters(model)

/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/huggingface_hub/file_download.py:1132: FutureWarning: `resume_download` is deprecated and will be removed in version 1.0.0. Downloads always resume when possible. If you want to force a new download, use `force_download=True`.
  warnings.warn(
Some weights of the model checkpoint at bert-base-uncased were not used when initializing BertForMaskedLM: ['bert.pooler.dense.bias', 'bert.pooler.dense.weight', 'cls.seq_relationship.bias', 'cls.seq_relationship.weight']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertFor

trainable params: 109514298 || all params: 109514298 || trainable%: 100.0


In [6]:
from adapters import ConfigUnion, LoRAConfig, PrefixTuningConfig, SeqBnConfig,SeqBnInvConfig,AdapterConfig,LoRAConfig

union_config = ConfigUnion(
    SeqBnInvConfig(reduction_factor=2),
    LoRAConfig(r=8, alpha=16),
)
model.add_adapter(adapter_name, config=union_config)

model.train_adapter([adapter_name])
model.active_adapters = adapter_name
print(model.adapter_summary())
fn.print_trainable_parameters(model)

Name                     Architecture         #Param      %Param  Active   Train
--------------------------------------------------------------------------------
mlm_union_F              union             7,682,688       7.055       1       1
--------------------------------------------------------------------------------
Full model                               108,891,648     100.000               0
trainable params: 8274816 || all params: 117196986 || trainable%: 7.060604783812444


In [7]:
from transformers import AutoTokenizer

reload(processed)
tokenizer = AutoTokenizer.from_pretrained(config.Config.TOKENIZER_NAME)

tokenized_dataset= processed.tokenize_dataset(train_target,tokenizer)
tokenized_dataset

Map:   0%|          | 0/77348 [00:00<?, ? examples/s]

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids'],
    num_rows: 77348
})

In [8]:
chunk_size = 128
# Slicing produces a list of lists for each feature
tokenized_samples = tokenized_dataset[444:470]

results = fn.group_texts(tokenized_samples, chunk_size)
for chunk in results["labels"]:
    print(f"'>>> Chunk length: {len(chunk)}'")

'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'
'>>> Chunk length: 128'


In [9]:
# we group texts and chunk them
lm_datasets = tokenized_dataset.map(fn.group_texts, batched=True,fn_kwargs={'chunk_size': chunk_size})
lm_datasets

Map:   0%|          | 0/77348 [00:00<?, ? examples/s]

Dataset({
    features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
    num_rows: 17714
})

In [10]:
from transformers import DataCollatorForLanguageModeling

data_collator = DataCollatorForLanguageModeling(tokenizer=tokenizer, mlm_probability=0.15)
samples = [lm_datasets[i] for i in range(2)]
for sample in samples:
    _ = sample.pop("word_ids")
dd=data_collator(samples)
for chunk in dd["input_ids"]:
    print(f"\n'>>> {tokenizer.decode(chunk)}'")


'>>> [CLS] have you any [MASK] to believe that [MASK] mistress was likely to take a sleeping powder [MASK] night [MASK] " [SEP] your mistress was [MASK] all night. [SEP] [CLS] do you want to see them? " [SEP] you can't [MASK] them. [SEP] [CLS] [MASK] groups your leader [MASK],'white said. [SEP] white said that was wherever your leader is. [SEP] [CLS] in the meantime [MASK] the young adventurers were sitting bolt upright, very events and ill at ease, in a taxi which, with [MASK] singular lack of originality commanded was [MASK] returning to the [MASK]tz via regent's park. [SEP] the adventurers walked home instead of taking [MASK] cab'

'>>> . [SEP] [CLS] my brother lawrence is convinced that we [MASK] making a fuss over nothing. [SEP] [MASK] brother is wrong when he thinks we are making a fuss [MASK] [SEP] [CLS] at eleven o'clock [MASK] tuppence made up her mind [MASK] [SEP] tuppence made a decision at [MASK] o'clock. [SEP] [CLS] a moment later he uttered a cry. [SEP] ten years later, 

In [11]:

downsampled_dataset = lm_datasets.train_test_split(
    test_size=0.1, seed=42
)
downsampled_dataset

DatasetDict({
    train: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 15942
    })
    test: Dataset({
        features: ['input_ids', 'token_type_ids', 'attention_mask', 'word_ids', 'labels'],
        num_rows: 1772
    })
})

In [12]:

reload(fn)
trainer = fn.train_mlm_model(model,adapter_name,data_collator,tokenizer, downsampled_dataset['train'],downsampled_dataset['test'])

huggingface/tokenizers: The current process just got forked, after parallelism has already been used. Disabling parallelism to avoid deadlocks...
To disable this warning, you can either:
	- Avoid using `tokenizers` before the fork if possible
	- Explicitly set the environment variable TOKENIZERS_PARALLELISM=(true | false)
/home/guest/.cache/pypoetry/virtualenvs/third-experments-xuKQSur9-py3.8/lib/python3.8/site-packages/accelerate/accelerator.py:446: FutureWarning: Passing the following arguments to `Accelerator` is deprecated and will be removed in version 1.0 of Accelerate: dict_keys(['dispatch_batches', 'split_batches', 'even_batches', 'use_seedable_sampler']). Please pass an `accelerate.DataLoaderConfiguration` instead: 
dataloader_config = DataLoaderConfiguration(dispatch_batches=None, split_batches=False, even_batches=True, use_seedable_sampler=True)
  warnings.warn(


In [13]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/56 [00:00<?, ?it/s]

>>> Perplexity: 14.24


In [14]:
trainer.train()

  0%|          | 0/9980 [00:00<?, ?it/s]

{'loss': 2.2011, 'grad_norm': 1.9283149242401123, 'learning_rate': 9.960000000000001e-05, 'epoch': 1.0}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.8401819467544556, 'eval_runtime': 3.0766, 'eval_samples_per_second': 575.957, 'eval_steps_per_second': 18.202, 'epoch': 1.0}
{'loss': 1.9027, 'grad_norm': 1.5654383897781372, 'learning_rate': 9.476793248945148e-05, 'epoch': 2.0}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.7565943002700806, 'eval_runtime': 3.0579, 'eval_samples_per_second': 579.485, 'eval_steps_per_second': 18.313, 'epoch': 2.0}
{'loss': 1.8074, 'grad_norm': 1.7456997632980347, 'learning_rate': 8.951476793248945e-05, 'epoch': 2.99}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.6767109632492065, 'eval_runtime': 5.4879, 'eval_samples_per_second': 322.893, 'eval_steps_per_second': 10.204, 'epoch': 3.0}
{'loss': 1.758, 'grad_norm': 1.7699341773986816, 'learning_rate': 8.426160337552744e-05, 'epoch': 3.99}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.661486029624939, 'eval_runtime': 5.2228, 'eval_samples_per_second': 339.282, 'eval_steps_per_second': 10.722, 'epoch': 4.0}
{'loss': 1.7336, 'grad_norm': 1.5519382953643799, 'learning_rate': 7.90084388185654e-05, 'epoch': 4.99}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.643464207649231, 'eval_runtime': 5.4836, 'eval_samples_per_second': 323.143, 'eval_steps_per_second': 10.212, 'epoch': 5.0}
{'loss': 1.6889, 'grad_norm': 1.7570252418518066, 'learning_rate': 7.375527426160338e-05, 'epoch': 5.99}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.6306347846984863, 'eval_runtime': 5.4752, 'eval_samples_per_second': 323.641, 'eval_steps_per_second': 10.228, 'epoch': 6.0}
{'loss': 1.6563, 'grad_norm': 1.6617778539657593, 'learning_rate': 6.850210970464134e-05, 'epoch': 6.99}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.602409839630127, 'eval_runtime': 5.4461, 'eval_samples_per_second': 325.371, 'eval_steps_per_second': 10.283, 'epoch': 7.0}
{'loss': 1.6542, 'grad_norm': 1.763481616973877, 'learning_rate': 6.324894514767933e-05, 'epoch': 7.98}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5913282632827759, 'eval_runtime': 5.364, 'eval_samples_per_second': 330.352, 'eval_steps_per_second': 10.44, 'epoch': 8.0}
{'loss': 1.629, 'grad_norm': 1.5569396018981934, 'learning_rate': 5.79957805907173e-05, 'epoch': 8.98}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5709609985351562, 'eval_runtime': 5.4722, 'eval_samples_per_second': 323.817, 'eval_steps_per_second': 10.233, 'epoch': 9.0}
{'loss': 1.6178, 'grad_norm': 1.6488170623779297, 'learning_rate': 5.2742616033755275e-05, 'epoch': 9.98}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5733543634414673, 'eval_runtime': 5.4512, 'eval_samples_per_second': 325.065, 'eval_steps_per_second': 10.273, 'epoch': 10.0}
{'loss': 1.5966, 'grad_norm': 1.6305493116378784, 'learning_rate': 4.748945147679325e-05, 'epoch': 10.98}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5742383003234863, 'eval_runtime': 5.4643, 'eval_samples_per_second': 324.287, 'eval_steps_per_second': 10.248, 'epoch': 11.0}
{'loss': 1.5901, 'grad_norm': 1.6779903173446655, 'learning_rate': 4.2236286919831224e-05, 'epoch': 11.98}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5368688106536865, 'eval_runtime': 7.9772, 'eval_samples_per_second': 222.133, 'eval_steps_per_second': 7.02, 'epoch': 12.0}
{'loss': 1.5795, 'grad_norm': 1.6414222717285156, 'learning_rate': 3.699367088607595e-05, 'epoch': 12.97}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5274773836135864, 'eval_runtime': 8.0246, 'eval_samples_per_second': 220.822, 'eval_steps_per_second': 6.979, 'epoch': 13.0}
{'loss': 1.5684, 'grad_norm': 1.706051230430603, 'learning_rate': 3.174050632911393e-05, 'epoch': 13.97}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5440263748168945, 'eval_runtime': 7.9404, 'eval_samples_per_second': 223.162, 'eval_steps_per_second': 7.053, 'epoch': 14.0}
{'loss': 1.5625, 'grad_norm': 1.7139652967453003, 'learning_rate': 2.6487341772151903e-05, 'epoch': 14.97}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.4904820919036865, 'eval_runtime': 7.9324, 'eval_samples_per_second': 223.389, 'eval_steps_per_second': 7.06, 'epoch': 15.0}
{'loss': 1.5534, 'grad_norm': 1.7226194143295288, 'learning_rate': 2.1234177215189873e-05, 'epoch': 15.97}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5237433910369873, 'eval_runtime': 7.9166, 'eval_samples_per_second': 223.834, 'eval_steps_per_second': 7.074, 'epoch': 16.0}
{'loss': 1.5501, 'grad_norm': 1.7135419845581055, 'learning_rate': 1.59915611814346e-05, 'epoch': 16.97}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5292073488235474, 'eval_runtime': 7.8662, 'eval_samples_per_second': 225.266, 'eval_steps_per_second': 7.119, 'epoch': 17.0}
{'loss': 1.5491, 'grad_norm': 1.7489233016967773, 'learning_rate': 1.0738396624472575e-05, 'epoch': 17.96}


  0%|          | 0/56 [00:00<?, ?it/s]

{'eval_loss': 1.5142015218734741, 'eval_runtime': 7.9118, 'eval_samples_per_second': 223.968, 'eval_steps_per_second': 7.078, 'epoch': 18.0}


Overwriting existing adapter 'mlm_union_F'.


{'train_runtime': 2307.1725, 'train_samples_per_second': 138.195, 'train_steps_per_second': 4.326, 'train_loss': 1.6774858522723073, 'epoch': 18.0}


TrainOutput(global_step=8982, training_loss=1.6774858522723073, metrics={'train_runtime': 2307.1725, 'train_samples_per_second': 138.195, 'train_steps_per_second': 4.326, 'train_loss': 1.6774858522723073, 'epoch': 18.0})

In [15]:
import math

eval_results = trainer.evaluate()

print(f">>> Perplexity: {math.exp(eval_results['eval_loss']):.2f}")

  0%|          | 0/56 [00:00<?, ?it/s]

>>> Perplexity: 4.62


In [16]:
import config.config as config

trainer.model.save_adapter(f"{config.Config.ADAPTER_SAVE_PATH}/{adapter_name}", adapter_name)


: 

#### This is a report of mlm for FICTION domain using whole genre as target and then splitt after tokenization with seed 42. I used computer metrix for preplexity and early stopping 

The results before:
>>> Perplexity: 14.24

The results after:
>>> Perplexity: 4.62